spacy_llm/pipeline/llm.py

```diff
@@ -237,28 +234,47 @@ def pipe(self, stream: Iterable[Doc], *, batch_size: int = 128) -> Iterator[Doc]
         """
         error_handler = self.get_error_handler()
         for doc_batch in spacy.util.minibatch(stream, batch_size):
-            is_cached = [doc in self._cache for doc in doc_batch]
-            noncached_doc_batch = [
-                doc for doc, cached_doc in zip(doc_batch, is_cached) if not cached_doc
-            ]
             try:
-                prompts = self._task.generate_prompts(noncached_doc_batch)
-                responses = self._backend(prompts)
-                modified_docs = iter(
-                    self._task.parse_responses(noncached_doc_batch, responses)
-                )
-                for doc, cached_doc in zip(doc_batch, is_cached):
-                    if cached_doc:
-                        doc = self._cache[doc]
-                        assert isinstance(doc, Doc)
-                        yield doc
-                    else:
-                        doc = next(modified_docs)
-                        self._cache.add(doc)
-                        yield doc
+                yield from iter(self._process_docs(doc_batch))
             except Exception as e:
                 error_handler(self._name, self, doc_batch, e)
 
+    def _process_docs(self, docs: List[Doc]) -> List[Doc]:
```

Comment:
I think this could technically be an `Iterable`?

```suggestion
    def _process_docs(self, docs: Iterable[Doc]) -> List[Doc]:
```